search/doc_vec.hpp

```diff
@@ -0,0 +1,322 @@
+  {
+    Init();
+  }
+
+  explicit QueryVec(Builder const & builder) : m_tws(builder.m_tws), m_prefix(builder.m_prefix)
+  {
+    Init();
+  }
+
+  // Computes cosine distance between |*this| and |rhs|.
+  double Similarity(DocVec<Token> const & rhs) const
+  {
+    size_t kInvalidIndex = std::numeric_limits<size_t>::max();
+
+    if (Empty() && rhs.Empty())
+      return 1.0;
+
+    if (Empty() || rhs.Empty())
+      return 0.0;
+
+    auto const & ls = m_tws;
+    auto const & rs = rhs.GetTokenWeightPairs();
+    auto const & maxWeights = rhs.GetMaxWeights();
+
+    ASSERT(std::is_sorted(ls.begin(), ls.end()), ());
+    ASSERT(std::is_sorted(rs.begin(), rs.end()), ());
+
+    std::vector<size_t> rsMatchTo(rs.size(), kInvalidIndex);
+
+    size_t i = 0, j = 0;
+    double dot = 0;
+    while (i < ls.size() && j < rs.size())
+    {
+      if (ls[i].m_token < rs[j].m_token)
+      {
+        ++i;
+      }
+      else if (ls[i].m_token > rs[j].m_token)
+      {
+        ++j;
+      }
+      else
+      {
+        dot += ls[i].m_weight * rs[j].m_weight;
+        rsMatchTo[j] = i;
+        ++i;
+        ++j;
+      }
+    }
+
+    auto const ln = impl::SqrL2(ls, m_prefix);
+    auto const rn = impl::SqrL2(rs);
+
+    // This similarity metric assumes that prefix is not matched in the document.
+    double const similarityNoPrefix = ln > 0 && rn > 0 ? dot / sqrt(ln) / sqrt(rn) : 0;
+
+    if (!m_prefix)
+      return similarityNoPrefix;
+
+    double similarityWithPrefix = 0;
+
+    auto const & prefix = *m_prefix;
+
+    // Let's try to match prefix token with all tokens in the
+    // document, and compute the best cosine distance.
+    for (size_t j = 0; j < rs.size(); ++j)
+    {
+      auto const & tw = rs[j];
+      if (!strings::StartsWith(tw.m_token.begin(), tw.m_token.end(), prefix.m_token.begin(),
+                               prefix.m_token.end()))
+      {
+        continue;
+      }
+
+      auto const i = rsMatchTo[j];
+
+      double nom = 0;
+      double denom = 0;
+      if (i == kInvalidIndex)
+      {
+        // If this document token is not matched with full tokens in a
+        // query, we need to update it's weight in the cosine distance
+        // - so we need to update correspondingly dot product and
+        // vector norms of query and doc.
+        auto const w = maxWeights[j];
+        auto const l = std::max(0.0, ln - prefix.SqrWeight() + w * w);
+
+        nom = dot + w * tw.m_weight;
+        denom = sqrt(l) * sqrt(rn);
+      }
+      else
+      {
+        // If this document token is already matched with |i|-th full
+        // token in a query - we know that completion of the prefix
+        // token is the |i|-th query token. So we need to update
+        // correspondingly dot product and vector norm of the query.
+        auto const w = ls[i].m_weight + m_maxWeights[i];
+        auto const l = ln - ls[i].SqrWeight() - prefix.SqrWeight() + w * w;
+
+        nom = dot + (w - ls[i].m_weight) * tw.m_weight;
```

Comment:
```(w - ls[i].m_weight)``` is ```m_maxWeights[i]```

Reply:
I know, but this is more readable, IMHO.